cmd/kubeadm/app/util/config/cluster.go

```diff
@@ -181,23 +183,83 @@ func getNodeNameFromKubeletConfig(kubeconfigDir string) (string, error) {
 	return strings.TrimPrefix(cert.Subject.CommonName, constants.NodesUserPrefix), nil
 }
 
-// getAPIEndpoint returns the APIEndpoint for the current node
-func getAPIEndpoint(data map[string]string, nodeName string, apiEndpoint *kubeadmapi.APIEndpoint) error {
-	// gets the ClusterStatus from kubeadm-config
-	clusterStatus, err := UnmarshalClusterStatus(data)
+func getAPIEndpoint(client clientset.Interface, nodeName string, apiEndpoint *kubeadmapi.APIEndpoint) error {
+	return getAPIEndpointWithBackoff(client, nodeName, apiEndpoint, constants.StaticPodMirroringDefaultRetry)
+}
+
+func getAPIEndpointWithBackoff(client clientset.Interface, nodeName string, apiEndpoint *kubeadmapi.APIEndpoint, backoff wait.Backoff) error {
+	var err error
+	var errs []error
+
+	if err = getAPIEndpointFromPodAnnotation(client, nodeName, apiEndpoint, backoff); err == nil {
+		return nil
+	}
+	errs = append(errs, errors.WithMessagef(err, "could not retrieve API endpoints for node %q using pod annotations", nodeName))
+
+	// NB: this is a fallback when there is no annotation found in the API server pod that contains
+	//     the API endpoint, and so we fallback to reading the ClusterStatus struct present in the
+	//     kubeadm-config ConfigMap. This can happen for example, when performing the first
+	//     `kubeadm upgrade apply` and `kubeadm upgrade node` cycle on the whole cluster. This logic
+	//     will be removed when the cluster status struct is removed from the kubeadm-config ConfigMap.
+	if err = getAPIEndpointFromClusterStatus(client, nodeName, apiEndpoint); err == nil {
+		return nil
+	}
+	errs = append(errs, errors.WithMessagef(err, "could not retrieve API endpoints for node %q using cluster status", nodeName))
+
+	return errorsutil.NewAggregate(errs)
+}
+
+func getAPIEndpointFromPodAnnotation(client clientset.Interface, nodeName string, apiEndpoint *kubeadmapi.APIEndpoint, backoff wait.Backoff) error {
+	var rawAPIEndpoint string
+	var lastErr error
+	// Let's tolerate some unexpected transient failures from the API server or load balancers. Also, if
+	// static pods were not yet mirrored into the API server we want to wait for this propagation.
+	err := wait.ExponentialBackoff(backoff, func() (bool, error) {
+		rawAPIEndpoint, lastErr = getRawAPIEndpointFromPodAnnotationWithoutRetry(client, nodeName)
+		return lastErr == nil, nil
+	})
 	if err != nil {
 		return err
 	}
+	parsedAPIEndpoint, err := kubeadmapi.APIEndpointFromString(rawAPIEndpoint)
+	if err != nil {
+		return errors.Wrapf(err, "could not parse API endpoint for node %q", nodeName)
+	}
+	*apiEndpoint = parsedAPIEndpoint
+	return nil
+}
 
-	// gets the APIEndpoint for the current machine from the ClusterStatus
-	e, ok := clusterStatus.APIEndpoints[nodeName]
-	if !ok {
-		return errors.New("failed to get APIEndpoint information for this node")
+func getRawAPIEndpointFromPodAnnotationWithoutRetry(client clientset.Interface, nodeName string) (string, error) {
```

Comment:
do we need the `...WithoutRetry` here, given there is no other `getRawAPIEndpointFromPodAnnotation` variants?
i think i saw comments on this topic before.

Reply:
The way functions were splitted were to ease unit testing for each one:

* `getAPIEndpoint` just calls to `getAPIEndpointWithBackoff` with `constants.StaticPodMirroringDefaultRetry`: no explicit testing needed here.

* `getAPIEndpointWithBackoff`: first tries to retrieve the endpoint with `getAPIEndpointFromPodAnnotation`, if it's not present, try to use the cluster status with `getAPIEndpointFromClusterStatus`. Tests ensure that this behavior happens.

* `getAPIEndpointFromPodAnnotation`: tries to fetch the api endpoint from pod annotations with a backoff. Tests ensure that this behavior happens. Calls to `getAPIEndpointFromPodAnnotationWithoutRetry`.

* `getAPIEndpointFromPodAnnotationWithoutRetry`: tries to fetch the api endpoint from pod annotations without any kind of retry. Tests ensure that this behavior happens.

* `getAPIEndpointFromClusterStatus`: already existed.

Please, note that this will be very simplified when we remove the ClusterStatus. I tried to separate all concerns on this interim while we must fallback to the cluster status if pod annotations are missing.

Reply:
I see where you are going, but can't we merge `getAPIEndpointWithBackoff` and `getAPIEndpoint`?

Reply:
in the code base we already have the "WithRetry" naming, but not "WithoutRetry"
it seems a bit odd that we now have a default that always retries.

Reply:
Requested here: https://github.com/kubernetes/kubernetes/pull/87656#discussion_r373976594